Naisten-Linja/naisten-linja-services
# Refactor local development environment setup

- As the project was setup with a backend node typescript along side a create-react-app, during development, typescript type check error would be displayed on the frontend due to the inclusion of backend source code in `tsconfig.json` (to provide correct editor linting support for both frontend and backend). Without this, the editor's TypeScript support would likely ignore configurations for backend codes. This PR aims to fix that issue by pointing `react-scripts` to use `tsconfig.frontend.json`, and use `tsconfig.json` as the base configuration (which is also used by the editor)
- The database docker run script is now included as a part of `npm run dev`, so there is no need to start the database separately.
- Update README and `.envrc.example`
- Add `FRONTEND_PORT` env variable instead of hard code port 3000 in the npm script.
- Rename `HOSTNAME` env variable to `APP_DOMAIN`, as the global `HOSTNAME` on your machines may get picked up by the node environment.  
- Add explicit eslint configuration, which is used by both frontend and backend code.
- Upgrade TypeScript to version 4
- Fix eslint and ts errors and warnings both in frontend and backend code after the addition of eslint and ts version upgrade.
- Add `isAuthenticated` express middleware to check for correct auth information and user roles in backend routes.

## How to test

Run `npm i` to install latest dependencies. You'll probably need to add the new  `export FRONTEND_PORT=3000` variable to `.envrc`. Make sure the Other `PORT` (used for backend) is 9000. 

Then `npm run dev` should start the full dev environment.

Test if you are be able to login, receive letters, response to letters and managed users.
